Performance improvements for FV

lindsayad opened the pull request:
After #22130 there were some slow-downs in some functor-based simulations. This brings back all the losses in the flow over sphere example plus a tiny bit more. Changes:

- Cache last warehouse lookup for Dirichlet boundary conditions
- Don't do pre-init evaluations for INSFV because we use on-the-fly evals entirely

### Pre 22130 graph

![pre-porous](https://user-images.githubusercontent.com/10248304/214193828-d443d98a-3a1b-468b-8d89-5896749edbcc.svg)

### Post 22130 graph
![post-porous](https://user-images.githubusercontent.com/10248304/214193832-f1f4833d-53e4-4006-9e2e-112bfd4395a6.svg)

### With this PR graph
![cache-fi-and-dont-preinit](https://user-images.githubusercontent.com/10248304/214193833-71816a49-e6ab-4372-a27e-cf6a1a89c9b8.svg)

Closes #23220 #23162

lindsayad committed:
Cache Dirichlet BC lookups

Sometimes we do repeated Dirichlet BC lookups and querying the
warehouse is a bit slow

Refs #23220

lindsayad committed:
Disable qp-evaluations for INSFV simulations

We don't use pre-init'd data for INSFV. We only use functors

Refs #23220

lindsayad commented:
Lol apparently we still use it for initial conditions

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1313888/) on 4e3182c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23233/site)

This comment will be updated on new commits.

lindsayad committed:
Adapt test suite to optimizations

Some code changes as well
- Make more FV kernels functor based

lindsayad committed:
Request qp computations from qp objects

Other code changes:
- Make second time derivative calculations request first time
  derivative calculations otherwise the second time derivative
  calculations will be inaccurate

lindsayad committed:
Cache dof indices

lindsayad commented:
Lol, eeked one more second with the dof indices "optimization". Nice use of time :-/

Mixing length test now runs about 6% faster than it did pre the porous PR

### pre porous

![mixing-pre-porous](https://user-images.githubusercontent.com/10248304/214456236-bb2e3b80-8666-4648-8726-9972b54de1b1.svg)

### with this PR, no dof indices optimization

![mixing-optimized](https://user-images.githubusercontent.com/10248304/214456285-c7efee61-7eb4-4cac-a44b-f8815260b0da.svg)

### with this PR, with dof indices optimization

![mixing-optimized-cache-dof-indices](https://user-images.githubusercontent.com/10248304/214456288-ac47b50a-f395-4e1c-b2dd-10a51e5386e2.svg)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1316247/) on 3df331f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>93fd19</th>
      <th colspan=4>#23233 3df331</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.62%</td>
      <td>84.62%</td>
      <td>-0.00%</td>
      <td>89.09%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81970</td>
      <td>81989</td>
      <td>+19</td>
      <td>49</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14897</td>
      <td>14904</td>
      <td>+7</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23233/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23233/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>93fd19</th>
      <th colspan=4>#23233 3df331</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.79%</td>
      <td>85.76%</td>
      <td>-0.03%</td>
      <td>81.82%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9978</td>
      <td>10003</td>
      <td>+25</td>
      <td>36</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1653</td>
      <td>1661</td>
      <td>+8</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23233/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23233/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23233/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 89.09% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 81.82% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
nice. Is there a diff option for pprof graphs? Just curious

lindsayad commented:
> nice. Is there a diff option for pprof graphs? Just curious

There is. I've used it before a while ago

lindsayad committed:
Restore FVTimeKernel and add FVFunctorTimeKernel

moosebuild commented:
Job [Test install](https://civet.inl.gov/job/1317215/) on 692cfce : invalidated by @lindsayad

moosebuild commented:
Job [External app tests](https://civet.inl.gov/job/1317317/) on 19f0eb0 : invalidated by @lindsayad

lindsayad commented:
@grmnptr @GiudGiud this is ready for review

grmnptr commented:
Got it! Mine is 50% done, will try to finish today!

grmnptr commented on framework/include/variables/MooseVariableBase.h:
```diff
@@ -172,6 +172,13 @@ class MooseVariableBase : public MooseObject,
+   * variables, but finite volume variables may not pre-compute quadrature point data unless this
+   * API is called
+   */
+  virtual void requireQpComputations() {}
```
This is cool! Just a quick question: in what cases do we need this for FV at the moment? I was thinking about this lately, if we didn't need thee evaluations we could get rid of most of the `MooseVariableFVData`, right? Or maybe I am wrong here. I suppose when we evaluate with quadrature points we need these.

lindsayad replied:
The fully compressible finite volume kernels are currently using qp indexing

grmnptr commented on framework/include/variables/MooseVariableDataFV.h:
```diff
@@ -417,6 +431,8 @@ class MooseVariableDataFV : public MooseVariableDataBase<OutputType>, public Mes
   using MooseVariableDataBase<OutputType>::_nodal_value_dotdot_old;
   using MooseVariableDataBase<OutputType>::_required_vector_tags;
+
+  friend void Moose::initDofIndices<>(MooseVariableDataFV<OutputType> &, const Elem &);
```
Can this be inlined? Also, whenever I see friend functions in moose it is really restrictive (like `Controls` and `ControllableParameters` is this going to be similarly restrictive?

lindsayad replied:
`initDofIndices` is a template function defined slightly higher up in the diff. Since it is templated it is implicitly marked as `inline`. What do you mean by similarly restrictive? The template function has been declared as a friend of this class so it can access private/protected data members.

grmnptr commented on framework/src/variables/MooseVariableFV.C:
```diff
@@ -372,10 +374,9 @@ template <typename OutputType>
-  this->_dof_map.dof_indices(elem, dof_indices, _var_num);
-  mooseAssert(dof_indices.size() == 1, "Wrong size for dof indices");
-  OutputType value = (*this->_sys.currentSolution())(dof_indices[0]);
+  Moose::initDofIndices(const_cast<MooseVariableFV<OutputType> &>(*this), *elem);
```
This is a very nice catch, especially when we have gradient evaluations where we use the value so many times!

grmnptr commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -608,6 +608,13 @@ class MooseVariableFV : public MooseVariableField<OutputType>
   const FieldVariablePhiValue & _phi_neighbor;
   const FieldVariablePhiGradient & _grad_phi_neighbor;
 
+  /// A member for caching retrieved Dirichlet conditions
```
I think we should go into details here with the reason why we need this.

lindsayad replied:
Removed this cache variable in favor of non-mutable map constructed during `initialSetup`

grmnptr commented on framework/src/variables/MooseVariableFV.C:
```diff
@@ -438,16 +443,19 @@ MooseVariableFV<OutputType>::getDirichletBC(const FaceInfo & fi) const
   {
     mooseAssert(bcs[0], "The FVDirichletBC is null!");
-
+    _face_to_diri = std::make_pair(&fi, bcs[0]);
```
Just a few comments here so people who look at this guy know why we are caching!

lindsayad replied:
Removed this cache variable in favor of non-mutable map constructed during `initialSetup`

grmnptr commented on modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i:
```diff
@@ -23,9 +23,11 @@ rho = 1
 [Variables]
   [vel_x]
     type = INSFVVelocityVariable
+    qp_calculations = true
```
Is there a way to know if we need this or not? I mean we should provide a guidance for the users so they can decide. Or is this just to make sure the exodiff checks out?

lindsayad replied:
I think I went through and made sure that (common) qp-consumer objects mark requirements for qp-calculations (see https://github.com/idaholab/moose/pull/23233/files#r1099323717), so I believe these input file modifications can actually be removed. However, if someone has forgotten to mark a consumer as needing qp calculations, then this input file option could be used to accomplish the same purpose.

lindsayad replied:
I will think more about whether there is a fool-proof way to detect whether qp-computations are required in the library per @GiudGiud's suggestion (https://github.com/idaholab/moose/pull/23233#discussion_r1100863857), and if so, then I can remove the input file option totally

lindsayad replied:
Ok I was able to delete the `qp_computations` parameter

grmnptr reviewed (changes requested):
I like it! Just a few questions/comments.

GiudGiud commented on framework/src/auxkernels/ParsedAux.C:
```diff
@@ -57,10 +57,18 @@ ParsedAux::ParsedAux(const InputParameters & parameters)
-      variables += (i == 0 ? "" : ",") + getFieldVar("args", i)->name();
+    {
+      auto * const field_var = getFieldVar("args", i);
+      field_var->requireQpComputations();
```
what is happening now? are we not pre-computing the values?

lindsayad replied:
not by default for INSFV/PINSFV/WCNSFV

GiudGiud commented on framework/src/auxkernels/VectorMagnitudeAux.C:
```diff
@@ -32,6 +32,15 @@ VectorMagnitudeAux::VectorMagnitudeAux(const InputParameters & parameters)
 {
+  auto require_qp_calcs = [this](const auto & var_name)
+  {
+    if (isCoupled(var_name))
+      getFieldVar(var_name, 0)->requireQpComputations();
```
any change we can move all this to the base class for auxkernels?
we can check which functors are requested, and request the QpComputations

lindsayad replied:
This isn't a functor though in this derived class. And many aux kernels now have functor versions which do not need qp computations. You could be registering false positives for needing qp computations and incurring computational cost that is not needed if you just made all coupled variables in every aux kernel `requireQpComputations`

GiudGiud replied:
yeah I suppose I worded this wrong. 

We can check which finite volume variables / other objects needing explicit Qp computations are being queried through a non-functor interface. Then we can requireQpComputations in the base class ?

lindsayad replied:
Alright I've updated `Coupleable` such that if a finite volume variable is requested through `add.*CoupledVar` then we `requireQpComputations`. So someone wishing to not trigger pre-init'd qp calculations should use `addParam<MooseFunctorName>`

GiudGiud commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -273,7 +278,7 @@ NSFVAction::validParams()
       std::vector<MooseFunctorName>(),
       "The ambient temperature for each block in 'ambient_convection_blocks'.");
 
-  params.addParam<CoupledName>(
```
can you delete this typedef of CoupledName then

lindsayad replied:
Hmm there are still a decent number of places in (P)INSFV and consequently in the action where we use coupled variables

lindsayad replied:
Changing that could be more follow on work

GiudGiud reviewed (commented):
so if we now init FV data on Qps when requested, can we use FV variables in FE kernels?

lindsayad commented on framework/include/variables/MooseVariableFV.h:
```diff
@@ -608,6 +608,13 @@ class MooseVariableFV : public MooseVariableField<OutputType>
   const FieldVariablePhiGradient & _grad_phi_neighbor;
 
+  /// A member for caching retrieved Dirichlet conditions
+  mutable std::pair<const FaceInfo *, const FVDirichletBCBase *> _face_to_diri;
```
I thought about this and went with the current code mostly out of laziness. I will try your suggestion out

GiudGiud replied:
instead rather than as well?

lindsayad replied:
Done. Excellent suggestion!

lindsayad commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -240,10 +240,16 @@ Coupleable::getFEVar(const std::string & var_name, unsigned int comp) const
 Coupleable::getFieldVar(const std::string & var_name, unsigned int comp) const
 {
-  return getVarHelper<MooseVariableFieldBase>(var_name, comp);
+  return const_cast<Coupleable *>(this)->getFieldVar(var_name, comp);
```
I suppose since this is a one-liner I can just write out the code. But if people modify one (const/non-const) they will have to remember to modify the other

lindsayad committed:
Address Guillaume and Peter reviews

- Typo fixes
- Use map from boundary ID to Dirichlet BC for run-time queries
  We construct this map during `initialSetup` and then can avoid
  doing slow Warehouse queries during residual/Jacobian evaluation.
  Great suggestion by Peter
- Don't use const cast in getFieldVar

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>
Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -1415,12 +1415,27 @@ class Coupleable
+   * Extract pointer to a base coupled field variable. Could be either a finite volume or finite
+   * element variable
+   * @param var_name Name of parameter desired
+   * @param comp Component number of multiple coupled variables
```
This is not necessarily an array variable. People can pass a multiple standard field variables

lindsayad commented:
Oh wow @grmnptr such beautiful code suggestion. Converting to a map constructed during `initialSetup` and avoiding doing any "run-time" queries leads to lopping off another 7 seconds of simulation time from 93 seconds for the previous best of the twice refined mixing length channel test to a new best of 86 seconds!

![use-map](https://user-images.githubusercontent.com/10248304/217973049-bf6e105c-f8f7-4c3d-b3ab-111a8b01ef69.svg)

lindsayad committed:
Automatically do qp calcs when using add*CoupledVar

lindsayad committed:
Only use functors in (P)INSFV

This is in honor of our framework change to treat any variable
retrieved because of add*CoupledVar as needing qp-computations.
E.g. we don't want any qp-calculations in (P)INSFV so we want
to remove all uses of add*CoupledVar and replace with
add*Param<MooseFunctorName>

lindsayad commented:
After refactoring all of (P)INSFV to not use `addCoupledVar` but instead use `addParam<MooseFunctorName>` we are even faster (although this could just be statistical noise)

![post-review](https://user-images.githubusercontent.com/10248304/218899742-2af2d54b-de03-4ae6-b845-9c25c9d556ab.svg)

lindsayad committed:
Remove qp_calculations parameter

We can do this now that we trigger qp calculations automatically
if an object requests a finite volume variable via `add.*CoupledVar`

GiudGiud commented on modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C:
```diff
@@ -48,11 +44,12 @@ INSFVWallFunctionBC::computeStrongResidual()
-    velocity(1) = _v_var->getElemValue(&elem);
-  if (_w_var)
-    velocity(2) = _w_var->getElemValue(&elem);
+  Moose::ElemArg elem_arg{&elem, false};
```
you should retrieve the skewness argument from the variable?

GiudGiud replied:
I guess we dont have the variables anymore. You d have to query the variable from intialSetup and keep the skewness argument in a class member

lindsayad replied:
I think @grmnptr expressed interest in having this be controlled more on an operator level as opposed to a global level, but I could be mis-remembering

GiudGiud replied:
I ll wait on Peter but essentially I m at an approval point here

grmnptr replied:
Yes I'll do these once I get to the point where unify limiting/interpolation. I think it would be best to do it on a per-kernel basis.

GiudGiud commented on modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C:
```diff
@@ -44,34 +44,23 @@ INSFVMassAdvectionOutflowBC::INSFVMassAdvectionOutflowBC(const InputParameters &
-    v(1) = _v_var->getBoundaryFaceValue(*_face_info);
-  if (_w_var)
-    v(2) = _w_var->getBoundaryFaceValue(*_face_info);
+  const auto boundary_face = singleSidedFaceArg();
```
same?

lindsayad replied:
we didn't do this before. Now you want to add it now that I've removed the variables?

GiudGiud replied:
seems like we should have the option to be consistently using skewness-correction everywhere

lindsayad replied:
you can open an issue for it. Feature addition/control is not the aim of this PR

GiudGiud replied:
For sure, but we need to agree that we ll do skewness correction at the kernel level if we re removing access to the variable.

It makes sense to realize now that we re not doing skewness-correction in all the operators as expected

GiudGiud replied:
#23438

GiudGiud commented on modules/navier_stokes/src/fvkernels/INSFVMixingLengthScalarDiffusion.C:
```diff
@@ -67,16 +58,18 @@ INSFVMixingLengthScalarDiffusion::computeQpResidual()
   constexpr Real offset = 1e-15; // prevents explosion of sqrt(x) derivative to infinity
 
-  const auto & grad_u = _u_var->adGradSln(*_face_info);
+  auto face = makeCDFace(*_face_info);
```
so when we make an argument like this, it defaults to no skewness correction right?
but what if the variable has skewness-correction on, does that bring it back?

lindsayad replied:
No it doesn't. `adGradSln` takes a `correct_skewness` argument which defaults to `false`, so the change here preserves previous behavior

GiudGiud commented on modules/navier_stokes/src/fvkernels/INSFVMomentumPressure.C:
```diff
@@ -19,44 +19,33 @@ INSFVMomentumPressure::validParams()
       momentum_component,
       "The component of the momentum equation that this kernel applies to.");
+  params.addParam<bool>(
+      "correct_skewness", false, "Whether to correct for mesh skewness in face calculations.");
```
that works but do we want this on every kernel and on every variable?
